### PR TITLE
Implement ZIP-244 authorizing data commitment (auth_digest)

### PR DESCRIPTION
**Pull request by conradoplg**
## Motivation

ZIP-244 specifies a new digest called "authorizing data commitment" or "auth_digest" that binds to authorization data (signatures, proofs). This must be implemented for Nu5.

Part of https://github.com/ZcashFoundation/zebra/issues/2048

### Specifications

https://zips.z.cash/zip-0244
https://zips.z.cash/zip-0225#modifications-to-zip-244

### Designs

N/A

## Solution

This adds a new method in `Transaction` to compute the auth_digest. It uses librustzcash to compute it, just like we do for the transaction ID and sighash.

## Review

This is a requirement for ZIP-221 work, but since that's ongoing it's not urgent.

@dconnolly or @teor2345 might want to review, but I think anyone else can review it too.

### Reviewer Checklist

  - [x] Code implements Specs and Designs
  - [x] Tests for Expected Behaviour
  - [ ] Tests for Errors

## Follow Up Work

N/A

---
This change is [Reviewable](https://reviewable.io/reviews/zcashfoundation/zebra/2547)

### CODE REVIEW

**Comment by teor2345** [comment]
I'm happy to leave this review to @dconnolly.

File: zebra-chain/src/block/merkle.rs
Comment (teor2345) [inline zebra-chain/src/block/merkle.rs]: Do we need to check for duplicate auth digests here, like we do in the transaction ID merkle tree?

Comment (teor2345) [reply zebra-chain/src/block/merkle.rs]: I just checked the spec, and we pad with zeroes, rather than duplicate hashes:
https://zips.z.cash/zip-0244#block-header-changes

File: zebra-chain/src/block/merkle.rs
Comment (teor2345) [inline zebra-chain/src/block/merkle.rs]: The transaction ID root pads with the previous transaction ID, but the auth data hash pads with zeroes.
Let's add a test case for this as well?

```rust
                // the root of a binary Merkle tree of transaction authorizing data commitments...
                // padded with leaves having the "null" hash value [0u8; 32]
                // https://zips.z.cash/zip-0244#block-header-changes
                [h1] => hash(h1, [0u8; 32]),
```

Comment (teor2345) [reply zebra-chain/src/block/merkle.rs]: I just noticed that this merkle tree code is shared between transaction IDs and the auth data root.

We could provide a closure that takes the previous hash, and returns the padding value?
(Or just duplicate the function?)

**Review by teor2345 — CHANGES_REQUESTED** [review]
We need to pad auth data merkle trees with zeroes:
https://zips.z.cash/zip-0244#block-header-changes

(The transaction ID merkle tree pads with the previous transaction ID.)

File: zebra-chain/src/block.rs
Comment (dconnolly) [inline zebra-chain/src/block.rs:208]: Nice 👍

File: zebra-chain/src/block/merkle.rs
Comment (dconnolly) [inline zebra-chain/src/block/merkle.rs:19]: 👍

File: zebra-chain/src/block/merkle.rs
Comment (dconnolly) [inline zebra-chain/src/block/merkle.rs]:
```suggestion
/// Compute the root of a Merkle tree as used in Bitcoin.
```

File: zebra-chain/src/block/merkle.rs
Comment (dconnolly) [inline zebra-chain/src/block/merkle.rs]:
```suggestion
/// Compute the root of a Merkle tree of transactions as used in Bitcoin.
```

File: zebra-chain/src/block/merkle.rs
Comment (dconnolly) [inline zebra-chain/src/block/merkle.rs]:
```suggestion
/// `hashes` must contain the hashes of the tree leaves (transactions) in some form.
```

File: zebra-chain/src/block/merkle.rs
Comment (dconnolly) [inline zebra-chain/src/block/merkle.rs]: Why write the result to the first element of the input, rather than return it? Maybe [Iterator.fold()](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.fold) could be used here and return the accumulator.

File: zebra-chain/src/block/merkle.rs
Comment (dconnolly) [inline zebra-chain/src/block/merkle.rs]: If `root()` results a new result we can then have
```rust
Self(root(hashes))
```

File: zebra-chain/src/block/merkle.rs
Comment (dconnolly) [inline zebra-chain/src/block/merkle.rs:141]: 👍

File: zebra-chain/src/block/merkle.rs
Comment (dconnolly) [inline zebra-chain/src/block/merkle.rs:168]: 👍

File: zebra-chain/src/block/merkle.rs
Comment (dconnolly) [inline zebra-chain/src/block/merkle.rs]: Same here, and what @teor2345 said about the different paddings.

I would be fine with having two `root()` functions, something like `hash_merkle_root()` and `hash_auth_data_root()`, since they are themselves small. Maybe a general one that takes the padding function or param, and then each of those are just calls to the underlying function.

File: zebra-chain/src/block/merkle.rs
Comment (dconnolly) [inline zebra-chain/src/block/merkle.rs]:
```suggestion
            // No test vectors for now, so just check it computes without panicking
```

File: zebra-chain/src/primitives/zcash_primitives.rs
Comment (dconnolly) [inline zebra-chain/src/primitives/zcash_primitives.rs:133]: 👍

File: zebra-chain/src/primitives/zcash_primitives.rs
Comment (dconnolly) [inline zebra-chain/src/primitives/zcash_primitives.rs:141]: A personal style nit, you may choose to ignore:

```suggestion

    let digest_bytes: [u8; 32] = alt_tx
```

File: zebra-chain/src/primitives/zcash_primitives.rs
Comment (dconnolly) [inline zebra-chain/src/primitives/zcash_primitives.rs:147]:
```suggestion

    AuthDigest(digest_bytes)
```

File: zebra-chain/src/transaction.rs
Comment (dconnolly) [inline zebra-chain/src/transaction.rs:176]: So, I know in [ZIP-244](https://zips.z.cash/zip-0244#authorizing-data-commitment), it says that for <= V4, it says that, `a placeholder value consisting of 32 bytes of 0xFF is used in place of the authorizing data commitment. This is only used in the tree committed to by hashAuthDataRoot`. Should that placeholder value be returned here? @teor2345

File: zebra-chain/src/transaction/auth_digest.rs
Comment (dconnolly) [inline zebra-chain/src/transaction/auth_digest.rs]: Should the inner value be `pub` or just `pub(crate)`, or does it need to be `pub` at all?

File: zebra-chain/src/transaction/tests/vectors.rs
Comment (dconnolly) [inline zebra-chain/src/transaction/tests/vectors.rs:486]: 💖

File: zebra-chain/src/block/merkle.rs
Comment (dconnolly) [inline zebra-chain/src/block/merkle.rs]: Does the wrapped value need to be `pub`?

File: zebra-chain/src/block/merkle.rs
Comment (dconnolly) [inline zebra-chain/src/block/merkle.rs]:
```suggestion
            let _auth_digest = block.transactions.iter().collect::<AuthDataRoot>();
```

File: zebra-chain/src/transaction/auth_digest.rs
Comment (dconnolly) [reply zebra-chain/src/transaction/auth_digest.rs]: If this is just need for comparison, maybe `impl From<AuthDigest> for [u8; 32]`

Comment (dconnolly) [inline zebra-chain/src/transaction/auth_digest.rs]:
```suggestion
pub struct AuthDigest(pub(crate) [u8; 32]);
```

File: zebra-chain/src/block/merkle.rs
Comment (dconnolly) [inline zebra-chain/src/block/merkle.rs]:
```suggestion
pub struct AuthDataRoot(pub(crate) [u8; 32]);
```

**Review by dconnolly — CHANGES_REQUESTED** [review]
Mostly small things, but teor++ on making sure we getting the different paddings right between the transaction merkle tree and the auth digest merkle tree, however we want to do that

File: zebra-chain/src/transaction.rs
Comment (teor2345) [reply zebra-chain/src/transaction.rs:176]: We need to use the `auth_digest` as a field in wide transaction IDs (see ticket #2449).

So I'll need a method that returns `Option<AuthDigest>`.
But feel free to create a wrapper method that substitutes the placeholder for `None`.

File: zebra-chain/src/block/merkle.rs
Comment (conradoplg) [reply zebra-chain/src/block/merkle.rs]: Oops, apologies for missing that, should've read the ZIP more carefully.

Exactly, I extracted the function so that it could be called for both paths, since I thought they had the same logic.

I got it even more wrong then I realized, since there are more differences:

- The way the padded leaves are defined prevent us from using the same algorithm, since the txid tree can just clone non-leaf trees where needed, but we can't just use `[0u8;32]` for non-leaves
- It uses a different hash algorithm

For this reason I've de-extracted the function and reverted to computing directly inside `from_iter`, which two separate implementations for txid and auth_digest. Done in 9cb90c84

File: zebra-chain/src/block/merkle.rs
Comment (conradoplg) [reply zebra-chain/src/block/merkle.rs]: I extracted that from the code in `from_iter` but the API did became strange. In any case moved the code back to `try_from` in 9cb90c84 so this no longer applies

File: zebra-chain/src/block/merkle.rs
Comment (conradoplg) [reply zebra-chain/src/block/merkle.rs]: I reverted the logic to be inside `from_iter` in 9cb90c84 so this no longer applies

File: zebra-chain/src/block/merkle.rs
Comment (conradoplg) [reply zebra-chain/src/block/merkle.rs]: I just copied `Root` but it makes sense to use `pub(crate)`. Solved by accepting the suggestion below

File: zebra-chain/src/block/merkle.rs
Comment (conradoplg) [reply zebra-chain/src/block/merkle.rs]: The way it works (I realize now) prevents us from calling the same function, since the leaves are handled differently. (The txid tree duplicates nodes, while the auth tree used padded leaves)

So I'm using two different implementations in 9cb90c84

File: zebra-chain/src/transaction.rs
Comment (conradoplg) [reply zebra-chain/src/transaction.rs:176]: My reasoning was exactly that, to make sure we don't use the placeholder value where we shouldn't.

It seems that the risk of _not_ using the placeholder where we _should_ is smaller, since it's in one specific place. So I suggest leaving this as is.

File: zebra-chain/src/transaction/auth_digest.rs
Comment (conradoplg) [reply zebra-chain/src/transaction/auth_digest.rs]: I just followed `SigHash` but it does make sense. Solved by accepting suggestion below

**Review by conradoplg — COMMENTED** [review]
Fixed the AuthDataRoot computation and added some tests

File: zebra-chain/src/block/merkle.rs
Comment (teor2345) [reply zebra-chain/src/block/merkle.rs]: The ZIP was only changed recently :-)

File: zebra-chain/src/block/merkle.rs
Comment (teor2345) [inline zebra-chain/src/block/merkle.rs:191]: > Pad with enough leaves to make the tree full (a power of 2).

Hmm, could be expensive!

Future optimisation: if computing transaction IDs turns out to be expensive, we might want to cache powers-of-two subtrees of `00` and `FF` bytes. Then substitute them for all-empty or all-v4 subtrees.

As an alternative, we could avoid recalculating transaction IDs / auth data hashes by:
* using the pre-calculated block transaction IDs in `PreparedBlock` in more places, or pre-calculating them earlier
* pre-calculating block auth data hashes in `PreparedBlock`
* pre-calculating the mempool transaction IDs and auth data hashes in `UnminedTransaction` (I'll do this in #2449)

So I think we can just open a ticket for precalculated fixed-value trees, and wait and see if we need one or both. (Most transactions should be v5 after NU5 activation, so we might not need to precalculate partial v4 auth data trees.)

**Review by teor2345 — APPROVED** [review]
Looks good to me.

I have a suggestion about a follow-up optimisation ticket, but it's not blocking.

File: zebra-chain/src/block/merkle.rs
Comment (daira) [reply zebra-chain/src/block/merkle.rs]: Yup, in https://github.com/zcash/zips/commit/f0c438de9b59da6b4127537a88eff8fde507e3d8

File: zebra-chain/src/block/merkle.rs
Comment (conradoplg) [reply zebra-chain/src/block/merkle.rs:191]: I agree! Created in https://github.com/ZcashFoundation/zebra/issues/2632

**Review by conradoplg — COMMENTED** [review]
Waiting for @dconnolly's approval

**Review by dconnolly — APPROVED** [review]
Awesome!

**Pull request merged**
